src/replication.c

```diff
@@ -831,32 +1024,78 @@ ask_replica_status_all(spec_t *spec)
 	MTX_UNLOCK(&spec->rq_mtx);
 }
 
-/*
- * This function processes the status response received from replica
- * and update spec's healthy/degrade count
- */
+static void
+update_rebuild_info(spec_t *spec, mgmt_ack_t *ack_data)
+{
+
+	uint64_t data_len = 0;
+
+	MTX_LOCK(&spec->rq_mtx);
+
+	if (spec->rebuild_info == NULL)
+		spec->rebuild_info = (mgmt_ack_t *)malloc(sizeof (mgmt_ack_t) *
+		    spec->rebuild_replica_count);
+
+	memcpy(&spec->rebuild_info[spec->rebuild_info_ack_recvd], ack_data,
+	    sizeof (mgmt_ack_t));
+	spec->rebuild_info_ack_recvd++;
+
+	if (spec->rebuild_info_ack_recvd == spec->rebuild_replica_count) {
+		data_len = spec->rebuild_info_ack_recvd * sizeof (mgmt_ack_t);
+		start_rebuild(spec->rebuild_info, spec->master_replica, data_len);
+		spec->rebuild_info = NULL;
+		REPLICA_LOG("Rebuild triggered on Replica ip:%s port:%d "
+		    "state:%d\n", spec->master_replica->ip,
+		    spec->master_replica->port,
+		    spec->master_replica->state);
+	}
+
+	MTX_UNLOCK(&spec->rq_mtx);
+}
+
 static int
 update_replica_status(spec_t *spec, replica_t *replica)
 {
 	zrepl_status_ack_t *repl_status;
-	replica_state_t last_status;
+	replica_state_t last_state;
 
 	repl_status = (zrepl_status_ack_t *)replica->mgmt_io_resp_data;
 
+	REPLICA_ERRLOG("Replica ip:%s port:%d state:%d rebuild status:%d\n",
+	    replica->ip, replica->port, repl_status->state,
+	    repl_status->rebuild_status);
+
 	MTX_LOCK(&spec->rq_mtx);
 	MTX_LOCK(&replica->r_mtx);
-	last_status = replica->state;
+
+	last_state = replica->state;
 	replica->state = (replica_state_t) repl_status->state;
 	MTX_UNLOCK(&replica->r_mtx);
 
-	if(last_status != repl_status->state) {
+	if (last_state != repl_status->state) {
 		if (repl_status->state == ZVOL_STATUS_DEGRADED) {
 			spec->degraded_rcount++;
 			spec->healthy_rcount--;
 		} else if (repl_status->state == ZVOL_STATUS_HEALTHY) {
+			/* master_replica became healthy*/
 			spec->degraded_rcount--;
 			spec->healthy_rcount++;
+			spec->master_replica = NULL;
+			spec->rebuild_in_progress = false;
+			REPLICA_ERRLOG("Replica:%s port:%d marked healthy,"
+		    	    " seting master_replica to NULL\n",
+		    	    replica->ip, replica->port);
 		}
+		/* Trigger another rebuild if required */
+		update_volstate(spec);
+	} else if ((repl_status->state == ZVOL_STATUS_DEGRADED) &&
+	    (repl_status->rebuild_status == ZVOL_REBUILDING_FAILED) &&
```

Comment:
let us take rebuild failure and re-triggering the rebuild in next PR.

Reply:
Next PR, you mean remove this code from this PR ?

Reply:
There are few pending things in re-triggering rebuild like detecting that rebuild threads are got destroyed at replica. 
And, to give focus on triggering rebuild in this PR, I suggest to move re-triggering rebuild part to next PR.